File: samples/java/src/main/java/shop/Money.java
package shop;

/** An amount in minor units. Never use floating point for currency. */
public final class Money {
    public static final String CURRENCY = "GBP";

    private final int pence;
    private final String currency;

    public Money(int pence) {
        this(pence, CURRENCY);
    }

    public Money(int pence, String currency) {
        this.pence = pence;
        this.currency = currency;
    }

    public int getPence() {
        return pence;
    }

    public String getCurrency() {
        return currency;
    }

    /** Adds two amounts, refusing to mix currencies. */
    public Money add(Money other) {
        if (!other.currency.equals(this.currency)) {
            throw new IllegalArgumentException("cannot add " + other.currency + " to " + currency);
        }
        return new Money(this.pence + other.pence, this.currency);
    }

    /** Multiplies the amount, rounding to the nearest penny. */
    public Money scaled(double factor) {
        return new Money((int) Math.round(pence * factor), currency);
    }

    public double asDecimal() {
        return pence / 100.0;
    }
}

File: samples/go/money.go
// Package bookstore prices and places book orders.
//
// Money is held in integer minor units. Never use floats for currency.
package bookstore

import "fmt"

const Currency = "GBP"

// Money is an amount in minor units with a currency code.
type Money struct {
	Pence    int
	Currency string
}

// NewMoney builds an amount in the default currency.
func NewMoney(pence int) Money {
	return Money{Pence: pence, Currency: Currency}
}

// Add returns the sum, refusing to mix currencies.
func (m Money) Add(other Money) (Money, error) {
	if other.Currency != m.Currency {
		return Money{}, fmt.Errorf("cannot add %s to %s", other.Currency, m.Currency)
	}
	return Money{Pence: m.Pence + other.Pence, Currency: m.Currency}, nil
}

// Scaled multiplies the amount, rounding to the nearest penny.
func (m Money) Scaled(factor float64) Money {
	return Money{Pence: int(float64(m.Pence)*factor + 0.5), Currency: m.Currency}
}

// AsDecimal renders the amount in major units.
func (m Money) AsDecimal() float64 {
	return float64(m.Pence) / 100.0
}
